app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt

```diff
@@ -3,83 +3,271 @@ package compiler.intermediate_form
 import compiler.ast.Function
 import compiler.ast.Type
 import compiler.ast.Variable
-
-data class FunctionDetailsGenerator(
-    val depth: Int,
-    val vars: Map<Variable, Boolean>,
-    val parameters: List<Variable>,
-    val functionCFG: ControlFlowGraph,
-    val function: Function
-)
+import compiler.common.reference_collections.ReferenceHashMap
 
 data class FunctionCallIntermediateForm(
     val callGraph: ControlFlowGraph,
     val result: IntermediateFormTreeNode?
 )
 
+data class VariableDisplayInfo(
+    val depth: ULong, // of parent function in display
+    val offset: ULong // from rbp of parent function
+)
+
+enum class VariableLocationType {
+    MEMORY,
+    REGISTER
+}
+
 val FUNCTION_RESULT_REGISTER = Register()
+val HELPER_REGISTER = Register()
+val BASE_POINTER_REGISTER = Register()
+val STACK_POINTER_REGISTER = Register()
 
-fun generateCall(fdg: FunctionDetailsGenerator, args: List<IntermediateFormTreeNode>): FunctionCallIntermediateForm {
-    val cfgBuilder = ControlFlowGraphBuilder()
+data class FunctionDetailsGenerator(
+    val depth: ULong,
+    val variablesLocationTypes: Map<Variable, VariableLocationType>, // should contain parameters
+    val parameters: List<Variable>,
+    val functionCFG: ControlFlowGraph,
+    val function: Function,
+    val variablesDisplayInfo: MutableMap<String, VariableDisplayInfo>,
+    val displayAddress: MemoryAddress
```

Comment:
I have some questions about constructing `FunctionDetailsGenerator`. `Function.Parameter` are not `Variable`, so the key type should be `Any` or maybe `NamedNode`. Same for `parameters` - it should be `List<Function.Parameter>`. Also what exactly are `variablesDisplayInfo` and `functionCFG`?

Reply:
For each variable we need to know at which level it was defined and what offset from rbp it had then. I assumed we need some info about this in constructor so we can use it in `genRead` and `genWrite`